src/borrow.rs

```diff
@@ -0,0 +1,586 @@
+//! Types to safely create references into NumPy arrays
```

Comment:
This document is filled with excellent detail 👍

If I could suggest one final improvement, I think that we can rearrange the order of content just so that users who want to understand what the module does get that information presented to them at the top.

I think it's as simple as adding a short paragraph after this one liner just listing the uses (e.g. obtaining ndarray views, iteration etc.) and moving the `Examples` section above `Rationale`.

Reply:
I tried implement your suggestions, please have another look.

This PR is getting a bit overwhelming w.r.t. the amount of comments accumulated, but I think there is also the still-open question at the end of https://github.com/PyO3/rust-numpy/pull/274#issuecomment-1071048638 to resolve.

Reply:
Looks great, I think we should proceed to merge! I think the doc is great now and I'm sure it will evolve further as users ask questions and the overlap resolution is implemented.

Reply:
Agreed, this is surely only the start of handling dynamic borrow checking of arrays and external users will surely bring an more expansive perspective on this. I think that at least, there should be no regressions for existing usage patterns involving `PyReadonlyArray` and the unsafe variant of `as_array_mut`.